cel/canonical_eventlog.go

```diff
@@ -117,25 +124,69 @@ type CEL struct {
 	Records []Record
 }
 
-// AppendEvent appends a new record to the CEL.
-func (c *CEL) AppendEvent(tpm io.ReadWriteCloser, pcr int, hashAlgos []crypto.Hash, event Content) error {
-	if len(hashAlgos) == 0 {
-		return fmt.Errorf("need to specify at least one hash algorithm")
-	}
+// generateDigestMap computes hashes with the given hash algos and the given event
+func generateDigestMap(hashAlgos []crypto.Hash, event Content) (map[crypto.Hash][]byte, error) {
 	digestsMap := make(map[crypto.Hash][]byte)
-
 	for _, hashAlgo := range hashAlgos {
 		digest, err := event.GenerateDigest(hashAlgo)
 		if err != nil {
-			return err
+			return digestsMap, err
 		}
 		digestsMap[hashAlgo] = digest
+	}
+	return digestsMap, nil
+}
 
-		tpm2Alg, err := tpm2.HashToAlgorithm(hashAlgo)
+// AppendEventRTMR appends a new RTMR record to the CEL. rtmrIndex indicates the RTMR to extend.
+// The index showing up in the record will be rtmrIndex + 1.
+func (c *CEL) AppendEventRTMR(client configfsi.Client, rtmrIndex int, event Content) error {
+	digestsMap, err := generateDigestMap([]crypto.Hash{crypto.SHA384}, event)
+	if err != nil {
+		return err
+	}
+
+	eventTlv, err := event.GetTLV()
+	if err != nil {
+		return err
+	}
+
+	err = rtmr.ExtendDigestClient(client, rtmrIndex, digestsMap[crypto.SHA384])
+	if err != nil {
+		return err
+	}
+
+	celrRTMR := Record{
+		RecNum:    uint64(len(c.Records)),
+		Index:     uint8(rtmrIndex) + 1,
+		Digests:   digestsMap,
+		Content:   eventTlv,
+		IndexType: RtmrTypeValue,
+	}
+
+	c.Records = append(c.Records, celrRTMR)
+	return nil
+}
+
+// AppendEvent appends a new PCR record to the CEL.
+//
+// Deprecated: Use AppendEventPCR
+func (c *CEL) AppendEvent(tpm io.ReadWriteCloser, pcr int, hashAlgos []crypto.Hash, event Content) error {
+	return c.AppendEventPCR(tpm, pcr, hashAlgos, event)
+}
+
+// AppendEventPCR appends a new PCR record to the CEL.
+func (c *CEL) AppendEventPCR(tpm io.ReadWriteCloser, pcr int, hashAlgos []crypto.Hash, event Content) error {
```

Comment:
I will again note that adding an AppendEvent[RegisterType] seems like the wrong interface. We will have to add a new one if we ever want to support NV registers, ARM registers, SVSM-simulated registers, GPU registers, etc.. This also introduces new dependencies every time we support a new register type since we have to create AppendEventARM and then import the ARM register library. Hopefully that's not often, but it's possible other companies decide to follow the Intel model.

I do think passing an extend func/interface as a parameter is the right call due to the decoupling, but I will leave this as optional.

Reply:
Agree, AppendEvent should work on arbitrary register type (instead of creating one AppendEvent for each). The change involves upstream go-tdx-guest, I will keep this for now, and update it in a later PR.